### PR TITLE
lodash - better return type for _.get with path in array form

### PR DESCRIPTION
**Pull request by WilliamChelman**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://lodash.com/docs/4.17.11#get
- [x] Increase the version number in the header if appropriate. => could not find version anywhere in source ?

With this, we can have something like
```typescript
const x = { a: { b: { c: { d: true } } } };
_.get(x, ["a", "b", "c", "d"]); // boolean | undefined
_.get(x, ["a", "b", "c2", "d"]); // any

type BS = {
  bool: boolean;
  s: string;
}

type NS = {
  n: number;
  s: string;
}

const y: { a: { b: BS | NS } } = {} as any;
_.get(y, ["a", "b"]); // BS | NS
_.get(y, ["a", "b", "bool"]); // any
_.get(y, ["a", "b", "n"]); // any
_.get(y, ["a", "b", "s"]); // string
```
with that we can have correct typing for path array of length up to 4, and it's still "any" for funkier scenarios.

### CODE REVIEW

**Review by aj-r — CHANGES_REQUESTED** [review]
Looks good. Please update the ts3 types as well. We have separate type definitions for ts >=3.1 now.

**Comment by WilliamChelman** [comment]
The failing doesn't seem to be linked to my changes, I expect that `npm run generate` was not called in the ts3.1/scripts folder in a past PR ? I'm not sure of what my options are here ? :stuck_out_tongue:

File: types/lodash/ts3.1/fp.d.ts [hunk 1]
```diff
@@ -2032,40 +2112,36 @@ declare namespace _ {
...
-    interface LodashMapValues3x1 {
+    interface LodashMapValues2x1 {
         <T>(obj: lodash.Dictionary<T> | lodash.NumericDictionary<T> | null | undefined): lodash.Dictionary<boolean>;
         <T extends object>(obj: T | null | undefined): { [P in keyof T]: boolean };
     }
-    type LodashMapValues5x1<T, TKey extends keyof T> = (obj: lodash.Dictionary<T> | lodash.NumericDictionary<T> | null | undefined) => lodash.Dictionary<T[TKey]>;
-    interface LodashMapValues6x1 {
+    interface LodashMapValues2x2 {
+        (iteratee: object): lodash.Dictionary<boolean>;
+        <TKey extends keyof T>(iteratee: TKey): lodash.Dictionary<T[TKey]>;
```
Comment (aj-r) [inline types/lodash/ts3.1/fp.d.ts]: TS is complaining that T is not declared (which is valid). I'm not sure what's going on but it's probably a bug in the `fp` generator. Probably remove this overload for now, and try to find a way to fix the generator in a separate PR. (That way it's easier to ensure that the generator fix didn't break other things.)

**Review by aj-r — CHANGES_REQUESTED** [review]
One of the Travis failures are for this test:

```ts
 fp.mapValues(valueIterator)(dictionary); // $ExpectType Dictionary<boolean>
```

I'm a little surprised that it changed, though. I think the new version of `mapValues` doesn't work very well in `fp`. I see 2 options for fixing this:
a) Pass explicit generic types to the test to ensure that it selects the right overload (e.g. `fp.mapValues<?, ?>(valueIterator)(dictionary);`
b) Manually undo the changes to mapValues, and figure out a better solution in a separate PR.

I'm leaning towards (b).

**Review by aj-r — APPROVED** [review]

**Comment by WilliamChelman** [comment]
@aj-r I see that there is still the "Revision needed" label, is there something else I should do for this PR ? :slightly_smiling_face:

**Comment by aj-r** [comment]
I don't think so... probably a bug with typescript-bot. @sandersn?

**Comment by sandersn** [comment]
Yep, typescript-bot forgets to remove Revision Needed sometimes. I haven't had time to go read that code to figure out why.

**Comment by uniqueiniquity** [comment]
Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

**Comment by WilliamChelman** [comment]
Thanks @uniqueiniquity , it's my first PR in open source, a real personal milestone 😸

**Pull request merged**